### ISSUE TITLE
RFE: cryptsetup: measure information about chosen cryptsetup unlock slot and mechanism to some TPM2 NV index

### ISSUE THREAD

**Issue by poettering** [systemd/systemd#29877]
We should measure the chosen unlock slot cryptsetup uses (including info if fido2, tpm2, pkcs11, recovery key, …), so that remote attestation can determine if the system is in an OK state. 

We are out of PCRs though afaics. At least none of the existing PCRs really seems right for this I think.

But we can add an NV index in PCR mode for this. It's a bit more work, but is equally powerful. 

This is supposed to continue the discussion from #24503, but focusses only on the one item left from that wishlist.

**Comment by berrange** [comment]
> We are out of PCRs though afaics. At least none of the existing PCRs really seems right for this I think.

Is it not reasonable to measure to the same place that we measure the "Root file system volume encryption key", so slot 15 ?

**Comment by poettering** [comment]
> Is it not reasonable to measure to the same place that we measure the "Root file system volume encryption key", so slot 15 ?

Different usecase I'd say:

PCR 15 is supposed to encode the identity of the system, i.e. it should be a concoction of reasonable machine IDs, root fs IDs. or in other words IDs that stay stable across reboots as long as the system is "the same". The idea is that orchestrators can bind secrets to a specific value of a PCR and it they can only be unlocked on that specific system. So PCR 15 – in my PoV – has the primary usecase of putting together stable access policies.

I think the used slot/mechanism to unlock LUKS probably has a usecase more something in the area of remote attestation: you want to validate that everything it did is kinda in order, and the resulting PCR is usualy not part of an access policy.

Hence I'd really like to keep this separate. One I think makes sense to use for unlock policies the other one doesn't, hence they shouldn't be mixed into the same PCR.

One thing is key: it's always easy to put together an access policy that locks against multiple different PCRs/NV indexes. Thus if you want them combined splitting them up is fine. But the reverse is not true: if you only maintained them combined you cannot generate an access policy that locks against them separately.

hence I am pretty sure it should not be PCR 15, but something new separate.

**Comment by berrange** [comment]
Ok, yes, that reasoning makes sense to me.

**Comment by berrange** [comment]
One could say that any examination of the choice of fido2 vs tpm vs password, etc is really more about auditing actions that have taken place.  There are many possible actions that could ultimately undermine security of a CVM. One example is logging into root account on a TTY, since the virtual keyboard is not to be trusted. It is unlikely to sensible to try capture all of this kind information in PCR extends, as IMHO it just ends up re-inventnig the audit subsystem.

IOW, should we just log an audit record about it rather than do any PCR extend. 

When creating an attestation report with the TPM, it is possible to sign a hash of the audit log and then send both the attestation report & audit log off to the server for validation.

**Comment by poettering** [comment]
> IOW, should we just log an audit record about it rather than do any PCR extend.

hmm, does the linux audit subsystem actually protect its logs via a PCR extension?

**Comment by DemiMarie** [comment]
Two problems her:

1. Do confidential VMs have a trustworth vTPM?
2. Will this cause NVRAM wear on physical TPMs?

**Comment by poettering** [comment]
> Two problems her:
> 
>     1. Do confidential VMs have a trustworth vTPM?

There are different models. Azure provides you with a vTPM on confidential VMs, as I understand. But I am not a confidential computing guy, I can't really tell you what the deal is regarding establishing trust in it.

>     2. Will this cause NVRAM wear on physical TPMs?

it should not. a PCR implemented via an NV index is a weird beast. You register it with the NV index APIs for the TPM, and that allocates it persistently, but the actually value is never persisted but always reset on boot. So in theory at least you should have one write when allocating it on OS install, but no writes when it is updated nor on reboots. That said, who knows what actual TPM implementations do. Maybe they didn't get the memo that PCRs-via-NV-indexes don't actually means PCR values stored in NVRAM.

**Comment by berrange** [comment]
> > Two problems her:
> > ```
> > 1. Do confidential VMs have a trustworth vTPM?
> > ```
> 
> There are different models. Azure provides you with a vTPM on confidential VMs, as I understand. But I am not a confidential computing guy, I can't really tell you what the deal is regarding establishing trust in it.

My hope & expectation is that ultimately all confidential VMs built with SEV-SNP or TDX technology will provide a trustworthy vTPM. Various people in the community have been pushing very strongly to utilize TPM as a pre-existnig standard technology, as opposed to say Intel's TDX specific RTMR concept which is a reinvention of the TPM PCRs but with even fewer slots :-(

In both SEV-SNP/TDX cases, the vTPM impl would not be running on the (untrustworthy) host, but would instead be running in guest context, but more privileged than the guest OS / guest firmware, such that it can be part of the attested software base. eg in Linux KVM with SEV-SNP, a vTPM would be one of the features SVSM exposed to a guest. With TDX 1.5 partitioning you could achieve a similar-ish architecture but I've not looked at TDX is as much detail. 

TL;DR:  expect a vTPM and it to be attestable and protected from both host and guest OS/guest firmware.

One caveat is that we should not assume the vTPM will be persistent. In some CVM scenarios it is possible (even likely) that the vTPM will be ephemeral. ie its root seeds will be re-manufactured on each power on and NVRAM will be empty. 
This is because there is more complexity involved in preserving vTPM state, as it would require an early attestation to take place in SVSM to unlock state.

**Comment by DemiMarie** [comment]
> an early attestation to take place in SVSM to unlock state.

Confidential computing also needs to deal with replay attacks on storage.  This would require that the vTPM be backed by attestable storage.

**Comment by poettering** [comment]
I posted #33276 now, which implements something like this.
